### ISSUE TITLE
torch.fft.rfft bug

### ISSUE THREAD

**Issue by jjhuang-ca** [pytorch/pytorch#63152]
## 🐛 Bug

When performing zero padding followed by rfft on cuda, incorrect results can happen on the second operation when padded length is even for longer sequences.  This behavior is observed on cuda only, not cpu.

## To Reproduce

```
import torch
import torch.fft as fft
import numpy as np
from scipy.fft import rfft


if __name__ == '__main__':
    dtype = torch.float64
    for device in ['cuda', 'cpu']:
        print('\ndevice={}'.format(device))
        Ly = 1600
        y = torch.randn(1, 1, Ly, dtype=dtype, device=device)
        for Lx in [160, 161, 1600, 1601, 16000, 16001]:
            x = torch.randn(1, 1, Lx, dtype=dtype, device=device)
            x = torch.nn.functional.pad(x, [0, Ly-1])
            yy = torch.nn.functional.pad(y, [0, Lx-1])
            X = fft.rfft(x)
            Y = fft.rfft(yy)
            X_s = rfft(x[0, 0, :].cpu().numpy())
            Y_s = rfft(yy[0, 0, :].cpu().numpy())
            dif_x = np.max(np.abs(X_s - X[0, 0, :].cpu().numpy()))
            dif_y = np.max(np.abs(Y_s - Y[0, 0, :].cpu().numpy()))
            print('Lx={}: dif_x={}, dif_y={}'.format(Lx, dif_x, dif_y))

```
Output:
```
device=cuda
Lx=160: dif_x=6.080392413311918e-12, dif_y=1.9477723502559983e-11
Lx=161: dif_x=1.4273168140722167e-14, dif_y=3.972054645195637e-14
Lx=1600: dif_x=8.206403321718826e-14, dif_y=113.69819800575286
Lx=1601: dif_x=5.0242958677880804e-14, dif_y=4.832213032818602e-14
Lx=16000: dif_x=3.0511999048007806e-13, dif_y=121.9687795227097
Lx=16001: dif_x=2.3198975494145925e-13, dif_y=5.54951617168815e-14

device=cpu
Lx=160: dif_x=2.416106516409301e-14, dif_y=7.105427357601002e-14
Lx=161: dif_x=1.1916163935586911e-14, dif_y=4.266955553143359e-14
Lx=1600: dif_x=7.105982447430966e-14, dif_y=9.500188217416325e-14
Lx=1601: dif_x=4.4938668397781776e-14, dif_y=4.278033777464534e-14
Lx=16000: dif_x=2.5579538487363607e-13, dif_y=7.8482010346102e-14
Lx=16001: dif_x=2.744571917763205e-13, dif_y=7.791709393054201e-14

```

## Expected behavior

dif_y for Lx=1601 and 16001 for cuda run should be around 1e-14.

## Environment

Collecting environment information...
PyTorch version: 1.9.0+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.10

Python version: 3.7.6 (default, Jan  8 2020, 19:59:22)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-4.19.56-1.el7.x86_64-x86_64-with-debian-buster-sid
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB

Nvidia driver version: 418.40.04
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] pytorch-ignite==0.2.0
[pip3] pytorch-lightning==1.1.1
[pip3] torch==1.9.0+cu111
[pip3] torchaudio==0.9.0
[pip3] torchinfo==1.5.3
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.10.0+cu111
[conda] numpy                     1.19.5                   pypi_0    pypi
[conda] pytorch-ignite            0.2.0                    pypi_0    pypi
[conda] pytorch-lightning         1.1.1                    pypi_0    pypi
[conda] torch                     1.9.0+cu111              pypi_0    pypi
[conda] torchaudio                0.9.0                    pypi_0    pypi
[conda] torchinfo                 1.5.3                    pypi_0    pypi
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.10.0+cu111             pypi_0    pypi

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @ngimel @mruberry @peterbell10 @walterddr

**Comment by mruberry** [comment]
@peterbell10 -- yikes! Would you take a look?

**Comment by xwang233** [comment]
Can't reproduce this on the latest source build with `numpy==1.19.5, scipy==1.6.0`.

Can reproduce this on 1.9.0+cu111 pip wheel.

GPU is V100.

**Comment by peterbell10** [comment]
At first glance this seems a lot like  #58724, but I will investigate further.

**Comment by peterbell10** [comment]
I've confirmed that it's only an issue with CUDA 11.1, the shapes that have issues are the `multi_bluestein_fft` transforms, and that a valid workaround is
```
torch.backends.cuda.cufft_plan_cache.max_size = 0
```
All of that is consistent with #58724. There's not really anything we can do about it except maybe disable plan cacheing for the affected CUDA versions.

**Comment by jjhuang-ca** [comment]
I've seen the same issue with 1.8.1+cu111.  I didn't try other versions of CUDA yet.

**Comment by jjhuang-ca** [comment]
@peterbell10 The workaround suggested worked.  Thanks.  Does it have any negative speed implications?

**Comment by peterbell10** [comment]
It does slow down repeated transforms as the cuFFT plan needs to be recreated each time.

**Comment by mruberry** [comment]
> All of that is consistent with #58724. There's not really anything we can do about it except maybe disable plan cacheing for the affected CUDA versions.

Sounds like a plan to me. cc @ngimel

**Comment by peterbell10** [comment]
I think I can add a check for large prime factors and only disable in those cases. That should fix the correctness issues without harming normal performance.

**Comment by Jup11ter** [comment]
Hello! @peterbell10 
I would like to consult some questions about this bug.
Based on the reproducible codes in [#63152](https://github.com/pytorch/pytorch/issues/63152) and [#67140](https://github.com/pytorch/pytorch/issues/67140), I found that many fft APIs seem to have some output problems on the Cuda backend in PyTorch 1.9.0.

### 63152 case
I have observed a similar bug with `torch.fft.rfftn` and `torch.fft.rfft2` in 1.9.0 version.
The output of the following code is like this:
```
device=cuda
Lx=16000: dif_x=4.291326661516388e-13, dif_y=117.88432003965872
Lx=16001: dif_x=1.7806149921485808e-13, dif_y=6.07087979554246e-14                
                                                                                               
device=cpu                                     
Lx=16000: dif_x=2.6963201038669064e-13, dif_y=9.099386826154683e-14                            
Lx=16001: dif_x=1.8474111129762605e-13, dif_y=4.7664655742347644e-14  
```

However, in 1.10.0, the output turns out to be this, which seems to be an expected output:
```
device=cuda
Lx=16000: dif_x=3.4576501834057003e-13, dif_y=1.0687705978178971e-13
Lx=16001: dif_x=1.9118749672591252e-13, dif_y=6.28539212249071e-14

device=cpu
Lx=16000: dif_x=2.5618982671915015e-13, dif_y=8.196303721106322e-14
Lx=16001: dif_x=1.797546735911271e-13, dif_y=5.695433295429594e-14
```


```python
import torch
import torch.fft as fft
import numpy as np
from scipy.fft import rfft

func_cls=torch.fft.rfftn
# func_cls=torch.fft.rfft2


if __name__ == '__main__':
    dtype = torch.float64
    for device in ['cuda', 'cpu']:
        print('\ndevice={}'.format(device))
        Ly = 1600
        y = torch.randn(1, 1, Ly, dtype=dtype, device=device)
        for Lx in [16000, 16001]:
            x = torch.randn(1, 1, Lx, dtype=dtype, device=device)
            x = torch.nn.functional.pad(x, [0, Ly-1])
            yy = torch.nn.functional.pad(y, [0, Lx-1])
            X = func_cls(x)
            Y = func_cls(yy)
            X_s = rfft(x[0, 0, :].cpu().numpy())
            Y_s = rfft(yy[0, 0, :].cpu().numpy())
            dif_x = np.max(np.abs(X_s - X[0, 0, :].cpu().numpy()))
            dif_y = np.max(np.abs(Y_s - Y[0, 0, :].cpu().numpy()))
            print('Lx={}: dif_x={}, dif_y={}'.format(Lx, dif_x, dif_y))

```

### 67140 case
`torch.fft.{r,i,h}fft` also have some output bugs on Cuda. For the following codes:

```python
import numpy as np
import torch
from scipy.stats import norm
from torch.fft import rfft
torch.manual_seed(0)
# func_cls=torch.fft.rfft
# func_cls=torch.fft.ifft
func_cls=torch.fft.hfft

device = 'cuda'
# device = 'cpu'
pdf = torch.from_numpy(
    norm.pdf(np.linspace(-10, 10, 1000).reshape(-1,1))
).to(device)
n_iter = 50
n_full = pdf.shape[0] * n_iter - n_iter + 1
fft0 = func_cls(pdf, axis=0, n=n_full)
fft1 = func_cls(pdf, axis=0, n=n_full)
x1=np.abs(fft0.cpu()[:, 0])
x2=np.abs(fft1.cpu()[:, 0])
print(x1)
print(x2)
print(np.allclose(x1,x2))
```

In 1.9.0 version, the output x1 and x2 is not close, in fact, the output looks like these:
```
tensor([99.9000, 99.7009, 99.1045,  ..., 98.1131, 99.1045, 99.7009],                           
       dtype=torch.float64)                                                                    
tensor([7.6946e-23, 9.3982e-23, 1.1474e-22,  ..., 1.4004e-22, 1.1474e-22,                      
        9.3982e-23], dtype=torch.float64)                                                      
False   
```

But in 1.10.0 version (or use cpu in 1.9.0 version), x1 and x2 are consistent on all these 3 APIs.

### A few puzzles
**I want to confirm whether the root cause of the bug on `rfft2/rfftn/rfft/ifft/hfft` is the same as the bug on `torch.fft.rfft`.**
**Are their unexpected results on Cuda related to cuFFT bug?**
It seems like these bugs are also fixed by this [pr #63327](https://github.com/pytorch/pytorch/pull/63327).

Thanks a lot!

**Comment by peterbell10** [comment]
@Jup11ter `rfft` and `rfftn` call into the same underlying private functions (`torch._fft_r2c`) so it's not surprising to me that you could run into this bug by making one equivalent transform with each function. Also, 1.10 is the PyTorch version that has the cuFFT workaround so this makes perfect sense.

**Comment by Jup11ter** [comment]
> @Jup11ter `rfft` and `rfftn` call into the same underlying private functions (`torch._fft_r2c`) so it's not surprising to me that you could run into this bug by making one equivalent transform with each function. Also, 1.10 is the PyTorch version that has the cuFFT workaround so this makes perfect sense.

Thank you very much for your explanation!
Now I have a much clear understanding of this bug and these APIs.
